src/lfortran/semantics/ast_common_visitor.h

```diff
@@ -733,6 +733,10 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
 
     std::map<std::string, ASR::ttype_t*> implicit_dictionary;
     std::map<uint64_t, std::map<std::string, ASR::ttype_t*>> &implicit_mapping;
+
+    std::map<std::string, std::pair<bool,std::vector<ASR::expr_t*>>> common_block_dictionary;
+    std::map<std::string, ASR::symbol_t*> &common_variables_hash;
```

Comment:
I think it has to be:
```suggestion
    std::map<uint64_t, ASR::symbol_t*> &common_variables_hash;
```
And then you use the hash of the `Variable` symbol, that is the local variable that is part of the common block. Jus tusing `std::string` won't work, since you can have another function with the same variable name (but different hash), that is not in a common block. So the hash is unique, while the name is not.

Reply:
Ohh yes, got it!

Reply:
This has trouble:
During `common / block_1 / A`, for `A` COMMON module is not yet created and hence not added to the symbol table. Forcing us to put `nullptr` in a hash table. Now after creation of module, when we got to populate it with `StructType` and variables we don't have information about local variables to link this module in hash table. 

One approach I think we can go with is, `std::map<uint64_t, std::pair<std::string, ASR::symbol_t*>> common_variables_hash`. Here`std::string` is common_module_name that is present with us, and `ASR::symbol_t*` is corresponding symbol which will be kept `nullptr`.

Now in `create_common_modules`, once all modules are created we can iterate over hash table, `resolve_symbol()` for module from global scope and then populate `ASR::symbol_t*`. 

I think keeping `std::string` here will not create any trouble as it indicates common modules name and those will be unique.

Reply:
I would create the common module soon. So maybe create A variable, process all variables, create a common module, then update the hashtable. So you never need to create a nullptr in a hashtable.

Reply:
Ohh yes, we can create module just when we encounter it and populate accordingly. I will do the changes. Thanks for this, all these is really helpful.